tests/test_navigation_object.py

```diff
@@ -0,0 +1,135 @@
+import pathlib
+
+import pytest
+
+
+srcdir = pathlib.Path(__file__).absolute().parent / "examples"
+
+
+def create_context_nav_assert(expected):
+    """
+    Sphinx event that is connected to the app to be able to make assertions
+    about the `context` while building
+    """
+    def assert_context_nav_objects(app, pagename, templatename, context, doctree):
+        if pagename in expected:
+            nav_object = context["get_nav_object"]()
+            assert nav_object == expected[pagename]
+        elif pagename not in ["genindex", "search"]:
+            raise ValueError("page {} not checked".format(pagename))
+
+    return assert_context_nav_objects
+
+
+@pytest.mark.sphinx(srcdir=srcdir / "test-basic", buildername="html")
+def test_navigation_object(app, status, warning):
+
+    expected = dict(
```

Comment:
I think that this could be tested with a regression check, no? Might be cleaner than hard-coding in the structure by hand.

Reply:
You mentioned it before, but can you remind me again what you exactly mean with a "regression check" in this case?

A reason I actually like the hardcoded structure here, is that this clearly visualizes what kind of navigation objects / lists of dicts we are actually producing. And which makes it easier to reason about them, compared to looking at the code and trying to understand what it creates.

Now, if I add more tests (for the different options, collapse=True/False, maxdepth, etc), it might get a bit tedious, so should certainly need to see if I can deduplicate some of the expected results then, if keeping to this approach.

Reply:
As an example check out the MyST-NB docs. Here's a regression test:

https://github.com/ExecutableBookProject/MyST-NB/blob/master/tests/test_parser.py#L5

here we check that the doctree hasn't changed: https://github.com/ExecutableBookProject/MyST-NB/blob/master/tests/test_parser.py#L19

and here we explicitly have the XML that is expected: https://github.com/ExecutableBookProject/MyST-NB/blob/master/tests/test_parser/test_basic_run.xml

so you are still hard-coding it into the repository, just as a separate regression file. The first time you run the test, it will fail and the file will be *created*. The next time you run the test, then it will check whether the in-memory object is *exactly* the same as the file on-disk, and error if they are different

Reply:
Thanks for the links! Some questions though .. :)

So you are testing that the doctree (the structure of the parsed document in docutils nodes, right?) is correct, which certainly makes sense for myst, since it is a custom doc parser. But is that something we need to test here?

I suppose if we use something like this, we would regression check the actual html output? (since the theme operates on the level of converting the doctree to html)

Reply:
yeah - we don't have to test the doctrees, I was just using it as an example, I think that pytest-regressions lets you test a lot of different file formats (e.g. I think that we also use it to test JSON, HTML, etc)